heartbeat/scheduler/scheduler.go

```diff
@@ -83,52 +95,83 @@ type taskOverSignal struct {
 	cont  []task // continuation tasks to be executed by concurrently for job at hand
 }
 
+// Schedule defines an interface for getting the next scheduled runtime for a job
 type Schedule interface {
+	// Next returns the next time a scheduled event occurs after the given time
 	Next(now time.Time) (next time.Time)
+	// Returns true if this schedule type should run once immediately before checking Next.
+	// Cron tasks run at exact times so should set this to false.
+	RunOnInit() bool
 }
 
 var debugf = logp.MakeDebug("scheduler")
 
+// New creates a new Scheduler
 func New(limit uint) *Scheduler {
 	return NewWithLocation(limit, time.Local)
 }
 
+// NewWithLocation creates a new Scheduler using the given time zone.
 func NewWithLocation(limit uint, location *time.Location) *Scheduler {
 	stateInitial := statePreRunning
 	return &Scheduler{
 		limit:    limit,
 		location: location,
 
-		state:  atomic.MakeInt(stateInitial),
-		jobs:   nil,
-		active: 0,
+		state: atomic.MakeInt(stateInitial),
+		done:  make(chan int),
 
-		addCh:    make(chan *job),
-		rmCh:     make(chan *job),
-		finished: make(chan taskOverSignal),
+		activeJobs:         activeJobsGauge,
+		activeTasks:        activeTasksGauge,
+		waitingTasks:       waitingTasksGauge,
+		jobsMissedDeadline: jobsMissedDeadlineCounter,
 
-		done: make(chan struct{}),
-		wg:   sync.WaitGroup{},
+		throttler: throttler.NewThrottler(limit),
 	}
 }
 
+// Start the scheduler.
 func (s *Scheduler) Start() error {
-	if !s.transitionRunning() {
-		return errors.New("scheduler can only be stopped from a running state")
-	}
+	s.state.Store(stateRunning)
+	s.throttler.Start()
+
+	interval := time.Second * 10
+
+	// Missed deadline reporter
+	go func() {
+		t := time.NewTicker(interval)
+
+		// Counter used to check if we're missing more checks now than before
+		missedAtLastCheck := uint64(0)
+		for {
+			select {
+			case <-s.done:
+				t.Stop()
+				return
+			case <-t.C:
+				// Take a snapshot of items that have missed the deadline.
+				// We only need to mutex long enough to do this. Since maps are reference types we can
+				// do this through assignment
```

Comment:
do we miss code here?

Reply:
Oops! That comment is from a previous draft. Will delete.